[CUMULUS-852] documentation for running tasks in lambda vs docker (#902)
nasa/cumulus: docs/data-cookbooks/run-tasks-in-lambda-or-docker.md

```yaml
    QueueGranules:
      Type: Task
      Resource: ${QueueGranulesLambdaFunction.Arn}
      Next: StopStatus
```

Given it has been discovered this task can no longer run in AWS Lambda, it can be run on the Cumulus ECS cluster by including the following in `app/config.yml`:

```yaml
  ecs:
    instanceType: t2.large
    desiredInstances: 1
    services:
      QueueGranules:
        docker: true
        image: cumuluss/cumulus-ecs-task:1.2.5
        memory: 4000
        count: 1
        envs:
          AWS_DEFAULT_REGION:
            function: Fn::Sub
            value: '${AWS::Region}'
        commands:
          - cumulus-ecs-task
          - '--activityArn'
          - function: Ref
            value: QueueGranulesActivity
          - '--lambdaArn'
          - function: Ref
            value: QueueGranulesLambdaFunction

  activities:
    - name: QueueGranules
```

and then modifying the corresponding Step Function State Machine definition in a workflow YAML file:

```yaml
    QueueGranules:
      Type: Task
      Resource: ${QueueGranulesActivity}
      Next: StopStatus
```

> ## How do I name my resources? A Short Primer on Cumulus Resource Naming Conventions

Reviewer: ```suggestion
## How do I name my resources? A short primer on Cumulus resource naming conventions
```

The capitalization seemed inconsistent, but ignore if that was intentional

Cumulus deployments currently depend on the [`kes`](https://github.com/developmentseed/kes) cli tool. `kes` references values in `app/config.yml`, `iam/config.yml`, lambda and workflow `.yml` files to populate the cloudformation templates that are a part of the `@cumulus/deployment` package. `kes` generates final cloudformation files and uploads them to AWS Cloudformation, creating or updating AWS Cloudformation Stacks.

When deploying AWS Lambdas and AWS Activities as detailed above, note the following naming conventions:

* `QueueGranules` in `lambdas.yml`: When kes generates the final cloudformation file, it will define a resource called `QueueGranulesLambdaFunction`. A valid lambda function resource name, such as `QueueGranulesLambdaFunction`, must be used in `app/config.yml` when passed as a command argument to `cumulus-ecs-task`.
* `QueueGranules` in `app/config.yml`: The string `QueueGranules` exists twice in the `app/config.yml` above.

>     * The first occurence is as a key under `services`. This key provides a descriptive prefix when naming the corresponding ECS Service. It will be included in final cloudformation yaml as an `AWS::ECS::Service` resource with the name `QueueGranulesECSService`. This ECS Service name prefix (`QueueGranules` in this example) can be anything since the service is not referenced in any other part of the deployment at this time.

Reviewer: ```suggestion
    * The first occurrence is as a key under `services`. This key provides a descriptive prefix when naming the corresponding ECS Service. It will be included in final cloudformation YAML as an `AWS::ECS::Service` resource with the name `QueueGranulesECSService`. This ECS Service name prefix (`QueueGranules` in this example) can be anything since the service is not referenced in any other part of the deployment at this time.
```

>     * The second occurence of `QueueGranules` is as a value under `activities: - name:`. The value of `activites: - name:` can be anything, but it should match the prefix of the `--activityArn` command argument passed to `cumulus-ecs-task`.

Reviewer: ```suggestion
    * The second occurrence of `QueueGranules` is as a value under `activities: - name:`. The value of `activites: - name:` can be anything, but the `--activityArn` command argument passed to `cumulus-ecs-task` must use this value plus `Activity` as a suffix (`QueueGranulesActivity` in this example).
```

---
id: run-tasks-in-lambda-or-docker
title: Run Step Function Tasks in Lambda or Docker
hide_title: true
---

# Running Step Function Tasks in AWS Lambda or Docker

## Overview

[AWS Step Function Tasks](https://docs.aws.amazon.com/step-functions/latest/dg/concepts-tasks.html) can run tasks on [AWS Lambda](https://aws.amazon.com/lambda/) or on [AWS Elastic Container Service (ECS)](https://aws.amazon.com/ecs/) as a Docker container.

> Lambda provides serverless architecture, providing the best option for minimizing cost and server management. ECS provides the flexibility to execute arbitrary code on any AWS EC2 instance type. 

Reviewer: I know it's specified in greater detail below, but should we briefly mention that ECS gives you more resources? Something like "..., allowing you run tasks that need more resources"

Author: Added a bit here https://github.com/nasa/cumulus/pull/902/commits/fd330bcc774c347bc53fcfdc576cb3f2be85ebdc let me know what you think

> ## When to use Lambda

Reviewer: Super good explanation 💯

The `cumulus-ecs-task` container takes an AWS Lambda Amazon Resource Name (ARN) as an argument (see `--lambdaArn` in the example below). This ARN argument is defined at deployment time. The `cumulus-ecs-task` worker polls for new Step Function Activity Tasks. When a Step Function executes, the worker (container) picks up the activity task and runs the code contained in the lambda package defined on deployment.

## Example: Replacing AWS Lambda with a Docker container run on ECS

Whether using AWS Lambda or Step Functions Activities, there will be a lambda package defined in `lambdas.yml`:

```yaml
QueueGranules:
  handler: index.handler
  timeout: 900
  memory: 3008
  source: node_modules/@cumulus/queue-granules/dist/
  useMessageAdapter: true
```

> Given the `QueueGranules` lambda package is already used Step Function State Machine definition, the following will exist in a workflows yml file:

Reviewer: ```suggestion
Given the `QueueGranules` lambda package is already used in a Step Function State Machine definition, the following will exist in a workflow YAML file:
```

```yaml
    QueueGranules:
      Type: Task

>       Resource: ${QueueGranulesLambdaFunction.Arn}

Reviewer: I don't think this is relevant for this PR, but I'd love to highly somewhere that `QueueGranulesLambdaFunction` is a name automatically generated based on the keys in `lambdas.yml`. Where do you think the right place for that would be?

Author: https://github.com/nasa/cumulus/pull/902/commits/42f4602a04f52dc70971b851f3f58e015c4958c0

      Next: StopStatus
```

Given it has been discovered this task can no longer run in AWS Lambda, it can be run on the Cumulus ECS cluster by including the following in `app/config.yml`:

```yaml
  ecs:
    instanceType: t2.large
    desiredInstances: 1
    services:
      QueueGranules:
        docker: true

>         image: cumuluss/cumulus-ecs-task:1.2.5

Reviewer: I assume this is the case, but is `cumuluss/` the right owner name for our Docker images? I've seen this and wondered if it was a typo. But I guess it's not since deployments are using this value right now

Author: Ya it's strange. Even stranger is it might have been a typo at the time, I vaguely remember @scisco telling me this. Or it could be that `cumulus` was taken. But yes `cumuluss` with 2 `s`'s is the Docker owner we are using at this time.

        memory: 4000
        count: 1
        envs:
          AWS_DEFAULT_REGION:
            function: Fn::Sub
            value: '${AWS::Region}'
        commands:
          - cumulus-ecs-task
          - '--activityArn'
          - function: Ref
            value: QueueGranulesActivity
          - '--lambdaArn'
          - function: Ref
            value: QueueGranulesLambdaFunction

  activities:

>     - name: QueueGranules

Reviewer: So this name comes from the fact that the key under services is `QueueGranules`, correct? Or is this name arbitrary?

Author: I believe the activity name can be anything but should match the activityArn passed to the services definition above:
```

Author: Also added documentation about this in https://github.com/nasa/cumulus/pull/902/commits/42f4602a04f52dc70971b851f3f58e015c4958c0

```

> and then modifying the corresponding Step Function State Machine definition:

Reviewer: ```suggestion
and then modifying the corresponding Step Function State Machine definition in a workflow YAML file:
```

```yaml
    QueueGranules:
      Type: Task

>       Resource: ${QueueGranulesActivity}

Reviewer: This name is based on the `activities` config, correct? Can we maybe add a sentence mentioning the naming convention? I'd just like to clear up as many of these naming conventions as possible

Reviewer: I'm also open to the idea that there should be a separate docs page highlighting these naming conventions, which could be handled as a separate ticket.

Author: Agree, added some documentation in a recent commit to this document but might be worth another documentation page 😬 just around naming conventions

Reviewer: Awesome, yeah maybe we can file a ticket and reference this documentation as a start that can be moved elsewhere later
